File: scripts/update_header.pl
#!/usr/bin/perl
use strict;		      # 'strict' insists that all variables be declared
use diagnostics;	      # 'diagnostics' expands the cryptic warnings
undef $/; # undefines the separator. Can read one whole file in one scalar.

# Extract C++ function information from a .cpp file and
# use that to update the corresponding .h file.

require $ENV{HOME} . '/.xemacs/indent_block.pl';

MAIN: {

  if (scalar @ARGV < 2){
    print "Must have at least two files as input\n";
    exit(0);
  }
  
  my ($cpp_file, $header_file, $text, %cpp_map);

  $cpp_file    = $ARGV[0];
  $header_file = $ARGV[1];
  
  open(FILE, "<$cpp_file") || die "Cannot open $cpp_file\n";
  $text = <FILE>;
  close(FILE);

  &parse_cpp($text, \%cpp_map);

  # override header file name if available
  if ($text =~ /\n\#\s*include\s*\"(.*?)\".*?header file/){
    $header_file = $1;

    if ($cpp_file =~ /^(.*\/)/){
      $header_file = $1 . $header_file;
    }
  }
  
  open(FILE, "<$header_file") || die "Cannot open $header_file\n";
  $text = <FILE>;
  close(FILE);

  $text = &parse_h($text, \%cpp_map);
  
  open(FILE, ">$header_file") || die "Cannot open $header_file\n";
  print FILE "$text";
  close(FILE);

  print "$header_file updated\n";
}

sub key_sig{

  # Signature is the number of commas
  
  my $block = shift;
  $block =~ s/\{.*?$//sg;

  $block = $block . "\n";
  $block =~ s/\/\/.*?\n//g;

  $block =~ s/[^,]//g;

  return $block;
}

sub parse_cpp {

  my $text    = shift;
  my $cpp_map = shift;

  my @blocks = &extract_blocks($text);

  my $block;
  foreach $block (@blocks){

    $block =~ s/\/\*.*?\*\///sg; # get rid of C-style comments

    # Will match things like: void  *  myname::myfun ( double x, double y) const{
    next unless ( $block =~ /(\w+[\s\*\&]+\w+\:\:\w+)\s*(\(.*?\)\s*\w*\s*)\{/s );

    my $key = $1;
    my $fun = $1 . $2 . ";\n\n";

    $fun =~ s/\s*;/;/g;
    
    # rm namespace from fun declaration
    $key =~ s/(\w+::)//;  
    $key =~ s/\s+/ /g;

    $cpp_map->{$key} = $fun;

  }

}
 
sub parse_h {

  my $text    = shift;
  my $cpp_map = shift;

  # identify the namespace in the h class
  my $namesp;
  # Look at the last of all namespaces (this is a bit hackish)
  if ($text =~ /^.*\n\s*(class|struct)\s+(\w+)\s*:*.*?\{/s){
    $namesp = $2;
  }else{
    print "Can't identify the namespace!\n";
    exit(0);
  }

  my (%h_map, $key);
  
  my @blocks = &extract_blocks($text);

  # Overwrite function prototypes that changed in the cpp file
  my $block;
  foreach $block (@blocks){

    $block =~ s/\/\*.*?\*\///sg; # get rid of C-style comments

    my $is_static = ($block =~ /^\s*static\s+/);

    # Strip the static keyword for now
    my $static = "";
    if ($is_static){
      $block =~ s/^(\s*static\s+)//g;
      $static = $1;
    }
    
    # match things like: void  *  myfun ( double x, double y){
    if  ( $block =~ /(\w+[\s\*\&]+[\w]+)(\s*\(.*\))/s ){
      $key = $1;
      $key =~ s/\s+/ /g;
    }else{
      # no match, put back the static and go on to the next block
      $block = $static . $block;
      next;
    }
    
    $block = $static . $block;
    
    # We'll need this map later
    $h_map{$key} = $block; 
    
    # Skip functions not having a match in the cpp file
    next unless (exists $cpp_map->{$key});
    
    # Ignore functions with preset params (having the equal sign somehwere)
    next if ($block =~ /=/);

    # Overwrite a .h entry with the corresponding .cpp entry.

    #print "overwriting: $block\n";
    #print "with $cpp_map->{$key}\n";
    
    $block = $cpp_map->{$key};

    # Put back the static keyword
    if ($is_static){
      $block = "static " . $block;
    }
    
    # rm the namespace and indent
    $block =~ s/(\w+::)(\w+\s*\()/$2/;
    $block = &indent_block ($block);
    
  }

  # See what new functions were declared in the cpp file and are missing in
  # the h file
  my @new_blocks;
  @new_blocks = ();
  foreach $key (keys %$cpp_map){

    next if (exists $h_map{$key});

    my $new_block = $cpp_map->{$key};

    next unless ($new_block =~ /$namesp\:\:/ ); # must be same namespace

    # rm the namespace and indent
    $new_block =~ s/(\w+::)(\w+\s*\()/$2/;

    $new_block =~ s/^\s*/  /g;
    $new_block = &indent_block ($new_block);

    push (@new_blocks, $new_block);
    
  }

  my $new_chunk = "";
  if ( scalar (@new_blocks) != 0 ) {
    $new_chunk = join ("", @new_blocks);
  }
  
  $text = join ("", @blocks);

  # Append the new blocks under the very last private:
  if ($new_chunk !~ /^\s*$/){
    $text =~ s/^(.*private:\s*)(.*?)$/$1$new_chunk$2/sg;
  }

  $text =~ s/\s*$/\n/g;
  
  return $text;
}
  
sub extract_blocks {

  my $text = shift;

  $text =~ s/\r//g; 
  $text =~ s/\t/ /g; 

  # Insert a newline after these keywords
  $text =~ s/(public|private|protected)\s*:\s*([^\}])/$1:\n\n  $2/g;
  
  # If a line ends with comma (followed optionally by comments),
  # remove any empty lines following that line, so that later we don't
  # think the empty line separates two blocks.  Don't break the
  # indentation.
  
  while ( $text =~ /(^.*?)(,[ ]*)(\/\/[^\n]*?\n|\n)([ ]*\n\s*)(.*?)$/s ){

    my $before = $1 . $2 . $3;
    my $spaces = $4;
    my $after  = $5;

    $spaces =~ s/^.*\n//g; 
    $text = $before . $spaces . $after;
  }
  
  my $sep = "x8o8A8newline";
  $text =~ s/(\n[ \t]*\n)/$1$sep/g;

  my @blocks = split($sep, $text);
  

  return @blocks;
}

File: trunk/python/update_header.py
#!/usr/bin/python

import sys
import os
import re # Perl-style regular expressions

# If a function declaration changed in a cpp file, this script updates
# the corresponding declaration in the corresponding h file.
    
def get_namespace(text):

  # Identify the namespace/class name in the given text. Look at the
  # last of all namespaces if there's more than one.

  namespace = ""

  m = re.match(r"^\s*(?:class|struct|namespace)\s+(\w+)\s*:*.*?\{",
              text, re.S)
  if m:
      namespace = m.group(1)
      return namespace
  else:
    return ""
  
def balanced_parens(text, lparen, rparen):

    text = re.sub("//.*?($|\n)", "", text)

    lp = re.findall(lparen, text)
    rp = re.findall(rparen, text)

    return len(lp) == len(rp)

def len_woc(text):
  # Length after stripping the comments
  text = re.sub("//.*?($|\n)", "", text)
  return len(text)

def extract_blocks(text, lparen, rparen):

    # A block consists of several consecutive lines such
    # that the parentheses in each block are balanced.
    
    lines = text.split("\n")

    blocks = []
    block  = ""
    lcount = 0
    
    for line in lines:
        
        lcount = lcount + 1
        block  = block + line
        
        if lcount < len(lines):
            block = block + "\n"

        if balanced_parens(block, lparen, rparen):
            blocks.append(block)
            block = ""

    # last block
    if block != "":
        blocks.append(block)

    return blocks

def indent_block(text):

  # Indent a block of text by the opening paranthesis
  
  lines        = text.split("\n")
  indent_level = 0

  for count in range(len(lines)):

    if indent_level > 0:
      lines[count] = re.sub("^[ \t]*", "", lines[count])
      lines[count] = " " * indent_level + lines[count]

    p = re.match("^(.*?\()", lines[count])
    if indent_level == 0 and p:
      indent_level = len(p.group(1))
      
  text = "\n".join(lines)

  # Strip whitespaces after the last newline (those whitespaces are an artifact
  # which cause problems later)
  text = re.sub("[ ]*$", "", text)

  return text

def parse_cpp(cpp_text, namespace):

    # Identify the function declarations. Map each function name
    # to the set of all functions with that name, e.g.,
    # myfun --> { void namesp::myfun(double x), double namesp::myfun(int s) }
    # Strip the namespace 

    cpp_map = {} # Needed if there are multiple functions with same name
    
    blocks = extract_blocks(cpp_text, '\(', '\)')
    count  = 1 # Used to keep the order of blocks. Must be > 0.
    
    for block in blocks:

        p = re.match("^(\w[^\n]*?\s)" # everything up to the namespace
                     + namespace
                     + "::(\w+)(\s*\(.*\)).*?\{", # fun name, and up to {
                     block, re.S)

        if not p: continue
        
        fun_name = p.group(2)
        fun_decl = p.group(1) + p.group(2) + p.group(3)

        if not cpp_map.has_key(fun_name): cpp_map[fun_name] = {}
          
        cpp_map[fun_name][fun_decl] = count
        count = count + 1

        #print "In parse_cpp:", fun_name, fun_decl, count
        
    return cpp_map

def parse_update_h(h_text, cpp_map, namespace):

    # Update the h file by overwriting each block with the
    # corresponding block from the cpp file.

    # This routine needs serious cleanup, and the ability to handle
    # multiple functions in the h file with the same name
    # (may need to change the output format of extract_blocks()).
    
    h_blocks_arr = extract_blocks(h_text, '\(', '\)')

    # Copy h_blocks_arr into a map structure as besides elements
    # i and i +1 we'll also want to have elements in between
    h_blocks = {}
    count    = -1
    for h_block in h_blocks_arr:
      count = count + 1
      h_blocks[count] = h_blocks_arr[count]
      
    hsorted_keys = h_blocks.keys()
    hsorted_keys.sort()
    
    h_map    = {}
    done_cpp = {}
    
    for count in hsorted_keys:

        h_block = h_blocks[count]
        
        p = re.match("""
        ^(\s*(?:static|virtual)?\s*)  # leading spaces, static, virtual
        (\w[^\n]*?\s)                 # type 
        (\w+)                         # function name 
        (\s*\(.*\))                   # list of arguments
        (.*)$                         # newline, const, etc. 
        """, h_block, re.S | re.X)

        if not p: continue

        fun_name         = p.group(3)
        h_map[fun_name]  = h_block # Will have problems with multiple funs with same name

        if not cpp_map.has_key(fun_name): continue
        if re.search("=", h_block):       continue # skip functions with default args

        # Find the function in the cpp file with the same name and
        # with the closest length (in terms of number of characters)
        min_error_sig       = 1e+100
        best_cpp_block      = ""
        best_cpp_with_extra = ""
        for cpp_block in cpp_map[fun_name]:

          # Skip any cpp function used earlier
          if done_cpp.has_key(cpp_block): continue

          # Add the extra info present in the h file and not the cpp
          # file, like "virtual", "static", etc.
          cpp_with_extra = p.group(1) + cpp_block + p.group(5) 

          error_sig = abs(len_woc(h_block) - len_woc(cpp_with_extra))
          
          if error_sig < min_error_sig:
            min_error_sig       = error_sig
            best_cpp_block      = cpp_block
            best_cpp_with_extra = cpp_with_extra
            
        if best_cpp_with_extra == "": continue # could not find a match
        
        # We found the cpp function with the closest signature
        h_blocks[count] = best_cpp_with_extra
        h_blocks[count] = indent_block(h_blocks[count])
        
        # mark that we used this one
        done_cpp[best_cpp_block] = 1
          
        #print "\n-------\nOverwriting\n\"" + h_block + "\"\nwith\n"
        #print "\"" + h_blocks[count] + "\n\"\n"

    # For a given new function in the cpp file find the closest
    # function in the cpp file before it which is present in the h
    # file. Use that information to decide where to insert the new
    # function in the h file.
    closestIndexBefore = -1
    closestFunBefore   = ""

    for key in cpp_map:
      
      if h_map.has_key(key): continue
      
      for cpp_block in cpp_map[key]:

        if h_map.has_key(key): continue
        index = cpp_map[key][cpp_block]
        #print "found new: ", key, cpp_block, index
        for key2 in cpp_map:
          for cpp_block2 in cpp_map[key2]:
            if cpp_map[key2][cpp_block2] < index and \
               cpp_map[key2][cpp_block2] > closestIndexBefore and \
               h_map.has_key(key2):
              closestIndexBefore = cpp_map[key2][cpp_block2]
              closestFunBefore   = key2 

        #print "closest index is ", closestFunBefore, "--", closestIndexBefore
        
        h_map[key] = cpp_block
        
        if closestIndexBefore >= 0:
          
          for count in h_blocks.keys():
            if h_blocks[count] == h_map[closestFunBefore]:
              closestIndexBefore = count
              #print "Corresp index in the h file is: ", closestIndexBefore
              
          indexToAdd = closestIndexBefore + 0.1
          
        else:
          # The case when the new function is the first in the file
          indexToAdd = 0.1

        h_blocks[indexToAdd] = "  " + cpp_block + ";\n"
        #print "New index is ", indexToAdd, h_blocks[indexToAdd]
         
      # End adding new blocks to the h file map
      
    # Regenerate the header file after including information from the cpp file
    h_text = ""
    hsorted_keys = h_blocks.keys()
    hsorted_keys.sort()
    for count in hsorted_keys:
      #print "count is ", count
      #print "fun is ", h_blocks[count]
      h_text += h_blocks[count]
    if not re.search("\n\s*$", h_text): h_text = h_text + "\n" # newline at the end

    # If the above operation does not succeed, insert the new functions
    # at the beginning of the namespace or in private sections of the class.
    new_blocks = {}
    for fun_name in cpp_map:
      
      if h_map.has_key(key): continue
      
      for cpp_block in cpp_map[key]:
        new_blocks[cpp_map[key][cpp_block]] = cpp_block # Needed for sorting
        
    # See first if we can match the namespace
    p = re.match("""
    ^(
    \s*
    (?:|namespace\s*\w+\s*\{)
    .*?[\n]
    )           # end of group 1
    ([ \t]*)    # indentation level (group 2)
    (.*)$       # group 3, whatever is left
    """, h_text, re.S | re.X)
    
    if p:
      groups = [p.group(1), p.group(2), p.group(3)]
    else:
     
      # See if we can match a class/struct
      p = re.match("""
      ^(
      .*\n\s*
      (?:public:|private:|(?:class|struct)\s*\w+\s*\{)
      .*?[\n]
      )           # end of group 1
      ([ \t]*)    # indentation level (group 2)
      (.*)$       # group 3, whatever is left
      """, h_text, re.S | re.X)
      
      if p:
        groups = [p.group(1), p.group(2), p.group(3)]
        
      else:
        print "Could not find a place to append the new blocks to"
        sys.exit(1)
      
    new_chunk    = ""
    indent_level = groups[1]
    
    sorted_keys = new_blocks.keys()
    sorted_keys.sort()
    for num in sorted_keys:
      new_chunk = new_chunk + indent_level + indent_block(new_blocks[num]) + ";\n"
      
    if new_chunk == "":
      return h_text # Nothing else to do
   
    h_text = groups[0] + new_chunk + groups[1] + groups[2]
   
    return h_text
 
if __name__ == '__main__':

    if len(sys.argv) < 2:
        print "Usage:", sys.argv[0], "file.cpp"
        sys.exit(1)

    cpp_file = sys.argv[1]
    
    h_file = cpp_file;
    h_file = re.sub(r"\.cpp", ".h", h_file)

    if (not os.path.exists(h_file)) or (not os.path.isfile(h_file)):
      print "File", h_file, "does not exist"
      sys.exit(1)
    
    fh = open(cpp_file, 'r'); cpp_text = fh.read(); fh.close()
    fh = open(h_file,   'r'); h_text   = fh.read(); fh.close()

    # Need this to deal with pecularities in my cpp style.
    #cpp_text = re.sub("\n\s*\/\*\s+\*\/\s*\(", "(", cpp_text)
    #h_text   = re.sub("\n\s*\/\*\s+\*\/\s*\(", "(", h_text)
    
    # Identify all scopes (blocks between { and }). For each of those
    # scopes which is a namespace/class/struct, find all cpp files
    # with it and update that scope.
    
    scopes = extract_blocks(h_text, '\{', '\}')
    count    = -1
    for scope in scopes:

      count = count + 1
      namespace = get_namespace(scope)
      if namespace == "": continue
      #print "namespace is ", namespace
      
      cpp_map       = parse_cpp(cpp_text, namespace)
      scopes[count] = parse_update_h(scope, cpp_map, namespace)
      
    h_text = "".join(scopes)

    fh = open(h_file, 'w'); fh.write(h_text); fh.close()

    h_file = re.sub(r"^.*/", "", h_file)
    print "Updated ", h_file

      
